lib/init.c

```diff
@@ -11,24 +11,26 @@ int metal_init(const struct metal_init_params *params)
 {
 	int error = 0;
 
-	memset(&_metal, 0, sizeof(_metal));
+	if (atomic_fetch_add(&_metal.common.ref_count, 1) == 0) {
+		_metal.common.log_handler = params->log_handler;
+		_metal.common.log_level   = params->log_level;
 
-	_metal.common.log_handler   = params->log_handler;
-	_metal.common.log_level     = params->log_level;
+		metal_list_init(&_metal.common.bus_list);
+		metal_list_init(&_metal.common.generic_shmem_list);
+		metal_list_init(&_metal.common.generic_device_list);
 
-	metal_list_init(&_metal.common.bus_list);
-	metal_list_init(&_metal.common.generic_shmem_list);
-	metal_list_init(&_metal.common.generic_device_list);
-
-	error = metal_sys_init(params);
-	if (error)
-		return error;
+		error = metal_sys_init(params);
+		if (error) {
+			atomic_fetch_sub(&_metal.common.ref_count, 1);
+			return error;
+		}
+	}
 
 	return error;
 }
 
 void metal_finish(void)
 {
-	metal_sys_finish();
-	memset(&_metal, 0, sizeof(_metal));
+	if (atomic_fetch_sub(&_metal.common.ref_count, 1) == 1)
+		metal_sys_finish();
```

Comment:
The `metal_state _metal` depends on the architecture, I would prefer to keep the memset to avoid side effect.